packages/web3-utils/src/utils.js

```diff
@@ -216,24 +234,34 @@ var hexToUtf8 = function(hex) {
     return utf8.decode(str);
 };
 
-
 /**
  * Converts value to it's number representation
  *
  * @method hexToNumber
  * @param {String|Number|BN} value
- * @return {String}
+ * @return {Number|BN}
  */
 var hexToNumber = function (value) {
     if (!value) {
         return value;
     }
 
-    if (typeof value === 'string' && !isHexStrict(value)) {
-        throw new Error('Given value "'+value+'" is not a valid hex string.');
+    if (typeof value === "string" && !isHexStrict(value)) {
+        throw new Error(
+            'Given value "' + value + '" is not a valid hex string.'
+        );
+    }
+
+    const [negative, hexValue] = String(value).startsWith("-")
+        ? [true, value.slice(1)]
+        : [false, value];
+    const num = toBN(hexValue);
+
+    if (num > Number.MAX_SAFE_INTEGER) {
+        return negative ? -num : num;
     }
```

Comment:
Does this basic check not present in the `BN` library?

Reply:
Not sure what you mean, I just want to return `number` if value is less than `MAX_SAFE_INTEGER` and `bigint` otherwise, like in 4.x